src/pke/examples/simple-ckks-bootstrapping.cpp

```diff
@@ -0,0 +1,202 @@
+
+    parameters.SetScalingModSize(dcrtBits);
+    parameters.SetScalingTechnique(rescaleTech);
+    parameters.SetFirstModSize(firstMod);
+
+    /*  A4) Bootstrapping parameters.
+    * We set a budget for the number of levels we can consume in bootstrapping for encoding and decoding, respectively.
+    * Using larger numbers of levels reduces the complexity and number of rotation keys,
+    * but increases the depth required for bootstrapping.
+	* We must choose values smaller than ceil(log2(slots)).
+    */
+    std::vector<uint32_t> levelBudget = {2, 2};
+
+    // We approximate the number of levels bootstrapping will consume to help set our initial multiplicative depth.
+    uint32_t approxBootstrapDepth = 9;
+
+    /* We give the user the option of configuring values for an optimization algorithm in bootstrapping.
+    * Here, we specify the giant step for the baby-step-giant-step algorithm in linear transforms
+    * for encoding and decoding, respectively. Either choose this to be a power of 2
+    * or an exact divisor of the number of slots. Setting it to be 0 allows OpenFHE to choose
+    * the values automatically.
+    */
+    std::vector<uint32_t> bsgsDim = {0, 0};
+
+    /*  A5) Multiplicative depth.
+    * The goal of bootstrapping is to increase the number of available levels we have, or in other words,
+    * to dynamically increase the multiplicative depth. However, the bootstrapping procedure itself
+    * needs to consume a few levels to run. We compute the number of bootstrapping levels required
+    * using GetBootstrapDepth, and add it to levelsUsedBeforeBootstrap to set our initial multiplicative
+    * depth.
+    */
+    uint32_t levelsUsedBeforeBootstrap = 23;
+    usint depth =
+        levelsUsedBeforeBootstrap + FHECKKSRNS::GetBootstrapDepth(approxBootstrapDepth, levelBudget, secretKeyDist);
+    parameters.SetMultiplicativeDepth(depth);
+
+    // Generate crypto context.
+    CryptoContext<DCRTPoly> cryptoContext = GenCryptoContext(parameters);
+
+    // Enable features that you wish to use. Note, we must enable FHE to use bootstrapping.
+    cryptoContext->Enable(PKE);
+    cryptoContext->Enable(KEYSWITCH);
+    cryptoContext->Enable(LEVELEDSHE);
+    cryptoContext->Enable(ADVANCEDSHE);
+    cryptoContext->Enable(FHE);
+
+    usint ringDim  = cryptoContext->GetRingDimension();
+    usint numSlots = ringDim / 2;
+    std::cout << "CKKS scheme is using ring dimension " << ringDim << std::endl << std::endl;
+
+    // Step 2: Precomputations for bootstrapping
+    cryptoContext->EvalBootstrapSetup(levelBudget, bsgsDim, numSlots);
+
+    // Step 3: Key Generation
+    auto keyPair = cryptoContext->KeyGen();
+    cryptoContext->EvalMultKeyGen(keyPair.secretKey);
+    // Generate bootstrapping keys.
+    cryptoContext->EvalBootstrapKeyGen(keyPair.secretKey, numSlots);
+
+    // Step 4: Encoding and encryption of inputs
+    // Inputs
+    std::vector<std::complex<double>> x1 = {0.25, 0.5, 0.75, 1.0, 2.0, 3.0, 4.0, 5.0};
```

Comment:
We should work with a vector of real numbers rather than complex numbers

Reply:
Fixed